documentation/supported_plugins.md

```diff
@@ -209,3 +210,152 @@ config:
+          key: ssh_password
+```
+
+Here, if the specified data is `{"ssh_user": "foo_user", "ssh_password": "foo_password"}`, then
+the `user` and `password` keys are resolved to `foo_user` and `foo_password`.
+Conversely, if the specified data is `{}`, then the `user` and `password`
+keys are resolved to `nil`. Since the `user` and `password` keys are part of
+the SSH transport's config, Bolt will delete these keys from the parsed config.
+Thus when data is `{}`, the above example will be parsed as
+
+```yaml
+targets:
+  - target1.example.com
+    config:
+      ssh: {}
+```
+
+> If you run Bolt on a target with the above config, then Bolt will auto-load
+the target's SSH config (default behavior).
+
+#### puppet_connect_data.yaml
+
+If a `puppet_connect_data.yaml` file exists in your project root, then the
+`puppet_connect_data` plugin will parse its data from this file. Below is an
+example of a valid `puppet_connect_data.yaml` file:
+
+```yaml
+ssh_user: foo_user
+ssh_password: foo_password
+```
+
+Here, the parsed data is `{"ssh_user": "foo_user", "ssh_password": "foo_password"}`.
+
+In the above example, `ssh_password` is specified as plain-text. For better security,
+you can specify plugin-reference values for your keys. Take a look at the following
+example:
+
+```yaml
+ssh_user: foo_user
+ssh_password:
+  _plugin: env_var
+  var: SSH_PASSWORD
+```
+
+Here, the `ssh_password` is resolved by the `env_var` plugin.
+
+**Note:** As of this writing, `puppet_connect_data` does _not_ cache resolved plugin
+reference values. Thus, given a key that resolves to a plugin reference, if that same
+key's referenced multiple times in the inventory file, then the plugin reference is
+re-evaluated each time. For example, given the above `puppet_connect_data.yaml` file,
+an inventory like:
+
+```
+targets:
+  - target1.example.com
+    config:
+      ssh:
+        user:
+          _plugin: puppet_connect_data
+          key: ssh_user
+        password:
+          _plugin: puppet_connect_data
+          key: ssh_password
+  - target2.example.com
+    config:
+      ssh:
+        user:
+          _plugin: puppet_connect_data
+          key: ssh_user
+        password:
+          _plugin: puppet_connect_data
+          key: ssh_password
+```
+
+will result in the `ssh_password` reference being evaluated _twice_ (once for
+the first target and once for the second).
+
+We plan on adding caching support in a later iteration of the plugin. Please
+file an issue if you'd like us to expedite this process.
+
+**Note:** You do not have to specify auto-loaded, `puppet_connect_data`-referenced
+SSH config in your `puppet_connect_data.yaml` file. The `puppet_connect_data` plugin
+resolves all unspecified keys to `nil`, which tells Bolt to remove the corresponding
+config keys from the target's transport config (meaning SSH config will be auto-loaded
+if that was the previous, pre-`puppet_connect_data` plugin behavior).
+
+#### Stored connection data
```

Comment:
I dont think this detail belongs in the bolt docs.

Reply:
I wonder if this whole section belongs in experimental? https://github.com/lucywyman/bolt/blob/main/documentation/experimental_features.md

Reply:
Hmm, perhaps that could be shortened to defining what we consider to be "secret" data. Don't really have any opinions on that.

Re: experimental, not sure. Maybe it could go to  a special `Puppet Connect` section? Or be included in the Puppet Connect docs? Probably worth asking about during product check-in later today.